wire-runtime/src/main/java/com/squareup/wire/Message.java

```diff
@@ -160,46 +161,40 @@ protected void setBuilder(Builder builder) {
     return adapter.fromInt(value);
   }
 
+  /**
+   * Create a {@code byte[]} of the serialized form of the data.
+   * <p>
+   * Note: This is an inefficient means of serializing data. For best performance, call
+   * {@link Wire#adapter} and then {@link MessageAdapter2#write}.
+   */
   @SuppressWarnings("unchecked")
   public byte[] toByteArray() {
-    return WIRE.messageAdapter((Class<Message>) getClass()).toByteArray(this);
-  }
-
-  public void writeTo(byte[] output) {
-    writeTo(output, 0, output.length);
-  }
-
-  public void writeTo(byte[] output, int offset, int count) {
-    write(WireOutput.newInstance(output, offset, count));
-  }
-
-  @SuppressWarnings("unchecked")
-  private void write(WireOutput output) {
-    MessageAdapter<Message> adapter = WIRE.messageAdapter((Class<Message>) getClass());
+    Buffer buffer = new Buffer();
     try {
-      adapter.write(this, output);
+      WIRE.messageAdapter((Class<Message>) getClass()).write(this, buffer);
     } catch (IOException e) {
-      throw new RuntimeException(e);
+      throw new AssertionError(e); // No I/O writing into Buffer.
     }
+    return buffer.readByteArray();
   }
 
-  public void writeUnknownFieldMap(WireOutput output) throws IOException {
+  void writeUnknownFieldMap(WireOutput output) throws IOException {
     if (unknownFields != null) {
       unknownFields.write(output);
     }
   }
 
   @SuppressWarnings("unchecked")
-  public int getSerializedSize() {
```

Comment:
Would we still write code like that today though? That pre-dates Okio by quite a bit and I think we've come to know cogs as a design that's very far from normal use case.

Today we'd write that as:
```java
public ByteString toBytes() {
  Buffer buffer = new Buffer();
  buffer.writeByte(dirty ? 1 : 0);
  build().getBackingObject().write(buffer);
  return buffer.readByteString();
}
```
which is much easier to understand.